downstairs/src/extent_inner_raw_v2.rs

```diff
@@ -0,0 +1,1013 @@
+        flush_number: u64,
+        gen_number: u64,
+    ) -> Result<(), CrucibleError> {
+        assert_eq!(block_written.len(), self.block_count() as usize);
+
+        let mut buf = vec![];
+        for c in block_written.chunks(8) {
+            let mut v = 0;
+            for (i, w) in c.iter().enumerate() {
+                v |= (*w as u8) << i;
+            }
+            buf.push(v);
+        }
+
+        let d = OnDiskMeta {
+            dirty,
+            flush_number,
+            gen_number,
+            ext_version: EXTENT_META_RAW_V2,
+        };
+        let mut meta = [0u8; BLOCK_META_SIZE_BYTES as usize];
+        bincode::serialize_into(meta.as_mut_slice(), &d).unwrap();
+        buf.extend(meta);
+
+        let offset = self.block_written_array_offset();
+
+        pwrite_all(file.as_fd(), &buf, offset as i64).map_err(|e| {
+            CrucibleError::IoError(format!("writing metadata failed: {e}"))
+        })?;
+
+        Ok(())
+    }
+
+    /// Decodes the block written array from the given file
+    ///
+    /// The file descriptor offset is not changed by this function
+    fn get_block_written_array(
+        &self,
+        file: &File,
+    ) -> Result<Vec<bool>, CrucibleError> {
+        let mut buf = vec![0u8; self.block_written_array_size() as usize];
+        let offset = self.block_written_array_offset();
+        pread_all(file.as_fd(), &mut buf, offset as i64).map_err(|e| {
+            CrucibleError::IoError(format!(
+                "could not read active contexts: {e}"
+            ))
+        })?;
+
+        let mut block_written = vec![];
+        for bit in buf
+            .iter()
+            .flat_map(|b| (0..8).map(move |i| b & (1 << i)))
+            .take(self.block_count() as usize)
+        {
+            // Unpack bits from each byte
+            block_written.push(bit != 0);
+        }
+        assert_eq!(block_written.len(), self.block_count() as usize);
+        Ok(block_written)
+    }
+
+    /// Returns the starting point in the file for the given block
+    fn block_pos(&self, block: BlockOffset) -> u64 {
+        let bpr = self.blocks_per_record();
+        let record = block.0 / bpr;
+        let block = block.0 % bpr;
+        record * self.recordsize
+            + block * (self.block_size() + BLOCK_CONTEXT_SIZE_BYTES)
+    }
+
+    /// Returns the position of the given block's context
+    fn context_slot(&self, block: BlockOffset) -> u64 {
+        self.block_pos(block) + self.block_size()
+    }
+
+    /// Returns the number of blocks that fit into a ZFS recordsize
+    fn blocks_per_record(&self) -> u64 {
+        // Each block contains data and a single context slot
+        let bytes_per_block = self.block_size() + BLOCK_CONTEXT_SIZE_BYTES;
+        self.recordsize / bytes_per_block
+    }
+
+    /// Checks whether there is padding after the given block
+    fn has_padding_after(&self, block: BlockOffset) -> bool {
+        // No padding at the end of the file
```

Comment:
Is this `end of the file` the actual end of the file, or just the end of the data/context section?

Reply:
The end of the data blocks section; fixed